### PR TITLE
Megalit gameplay logic

### PR DESCRIPTION
**Pull request by KirbyG**

### CODE REVIEW

File: common_constants.py [hunk 1]
```diff
@@ -1,18 +1,100 @@
...
+          'border': (0, 0, 0),
+          'tip':  (110,  110, 110),
+          'gripped': (50, 50, 50)
+        }
+
+class Vector:
+    def __init__(self, x, y):
+        self.x = x
+        self.y = y
+        from math import sqrt
```
Comment (JacobLezberg) [inline common_constants.py:22]: I don't love the import statement inside init. I think we should move this to the top of the file

File: common_constants.py [hunk 2]
```diff
@@ -1,18 +1,100 @@
...
+        self.magnitude = sqrt(self.x * self.x + self.y * self.y)
+
+    def __matmul__(self, other):
+        return self.x * other.x + self.y * other.y
+
+    # pure function
```
Comment (JacobLezberg) [inline common_constants.py:28]: ...Meaning this is overriding an abstract method?

File: np-gravity.py [hunk 3]
```diff
@@ -87,7 +88,7 @@ def draw(screen, game):
 
 # create game instance of the correct type
 puzzle = Puzzle(raw_input)
-game = Megalit(puzzle) if args.megalit else Popils(puzzle)
+game = Megalit(puzzle) if not args.megalit else Popils(puzzle)
```
Comment (JacobLezberg) [inline np-gravity.py]:
```suggestion
game = Megalit(puzzle) if args.megalit else Popils(puzzle)
```
I think this was correct the old way. If I run np-gravity w/ no parameters right now, it builds a megalit level (but the window title says "Popils Reduction"

File: megalit.py [hunk 4]
```diff
@@ -1,5 +1,196 @@
-from game import Game
+# this file contains the Megalit class and its Slab helper class
 
+import builtins
```
Comment (JacobLezberg) [inline megalit.py]: What is this import needed for? Googling makes it seem like it's for Python 2/3 compatibility

File: np-gravity.py [hunk 5]
```diff
@@ -117,6 +118,8 @@ def draw(screen, game):
                     game.update(UP)
                 elif event.key == pygame.K_DOWN:
                     game.update(DOWN)
+                elif event.key == pygame.K_SPACE:
```
Comment (JacobLezberg) [inline np-gravity.py:121]: Is this to change grip in Megalit only?

File: README.md [hunk 6]
```diff
@@ -1 +1,23 @@
...
+    class breakouts
+    megalit
+        mechanics
+        reduction (fix obvious issue first)
+        solver
+
+TODO (both)
+    make code presentable
+    contact Professor Williams
+    write the actual paper
```
Comment (JacobLezberg) [inline README.md]:
```suggestion
We have developed a novel proof technique for demonstrating NP-completeness of simple arcade games that rely on gravity as a primary mechanic. This project supports an upcoming paper (<paper-filename>.pdf) that applies the technique to Popils and Megalit. 

3SAT Example Library: https://www.cs.ubc.ca/%7Ehoos/SATLIB/benchm.html

TODO (jacob)
    remove manual cmd line entry
    adopt DIMACS CNF input format
    add input example to this file
    convert existing files to new format and .cnf extension
    Make README more useful

TODO (kirby)
    blocks can see their slabs
    scrolling
    class breakouts
    megalit
        mechanics
        reduction (fix obvious issue first)
        solver

TODO (both)
    fix Game repr
    allow exit during solve
    refactor code (comments and organization)
    write the actual paper
```

**Review by JacobLezberg — CHANGES_REQUESTED** [review]

File: common_constants.py (hunk 2 above)
Comment (KirbyG) [reply common_constants.py:28]: no, pure in the functional programming sense of having no side effects

File: common_constants.py (hunk 1 above)
Comment (KirbyG) [reply common_constants.py:22]: done

File: np-gravity.py (hunk 3 above)
Comment (KirbyG) [reply np-gravity.py]: good catch, i just changed it for testing purposes

File: megalit.py (hunk 4 above)
Comment (KirbyG) [reply megalit.py]: i think vs code added it, but it also seems to be unused currently so i removed it

File: np-gravity.py (hunk 5 above)
Comment (KirbyG) [reply np-gravity.py:121]: correct. i believe it won't cause any issues in popils

**Review by JacobLezberg — APPROVED** [review]

**Pull request merged**